Expose document.fragmentDirective (feature detection for text fragments)

sideshowbarker opened the pull request:
#### 04bd634bf20976f70014303e66875f22f8d65946
<pre>
Expose document.fragmentDirective (feature detection for text fragments)
<a href="https://bugs.webkit.org/show_bug.cgi?id=273466">https://bugs.webkit.org/show_bug.cgi?id=273466</a>

Reviewed by Ryosuke Niwa.

This change makes WebKit expose the document.fragmentDirective property
from <a href="https://wicg.github.io/scroll-to-text-fragment/#feature-detectability">https://wicg.github.io/scroll-to-text-fragment/#feature-detectability</a> —
to enable developers to programmatically detect that WebKit supports the
scroll-to-text-fragment feature. (Controlled by the new preference
ScrollToTextFragmentFeatureDetectionEnabled, also introduced in this change.)

Otherwise, without this change, developers can’t programatically detect
that WebKit supports the scroll-to-text-fragment feature.

* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-api-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-api.html: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/dom/Document.h:
(WebCore::Document::fragmentDirectiveForBindings):
* Source/WebCore/dom/Document.idl:
* Source/WebCore/page/FragmentDirective.h: Added.
(WebCore::FragmentDirective::create):
* Source/WebCore/page/FragmentDirective.idl: Added.

Canonical link: <a href="https://commits.webkit.org/279258@main">https://commits.webkit.org/279258@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a5181e609d53fd13313e134b373fdd4aa4f35da7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52809 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32146 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5296 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56087 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3532 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55114 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/38855 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3256 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42869 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2290 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54907 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29865 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45591 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23984 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/27038 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2908 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1691 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/46163 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48826 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3059 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57680 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/52322 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27950 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3121 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50262 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29170 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45783 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49541 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11555 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30088 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/64626 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28924 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12264 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/page/FragmentDirective.h:
```diff
@@ -0,0 +1,45 @@
+
+class FragmentDirective : public RefCounted<FragmentDirective> {
+    public:
+        static Ref<FragmentDirective> create()
```
Nit: Wrong indentation. public/private shouldn't be indented.

sideshowbarker committed:
Expose document.fragmentDirective (feature detection for text fragments)
https://bugs.webkit.org/show_bug.cgi?id=273466

Reviewed by Ryosuke Niwa.

This change makes WebKit expose the document.fragmentDirective property
from https://wicg.github.io/scroll-to-text-fragment/#feature-detectability —
to enable developers to programmatically detect that WebKit supports the
scroll-to-text-fragment feature. (Controlled by the new preference
ScrollToTextFragmentFeatureDetectionEnabled, also introduced in this change.)

Otherwise, without this change, developers can’t programatically detect
that WebKit supports the scroll-to-text-fragment feature.

* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-api-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-api.html: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/dom/Document.h:
(WebCore::Document::fragmentDirectiveForBindings):
* Source/WebCore/dom/Document.idl:
* Source/WebCore/page/FragmentDirective.h: Added.
(WebCore::FragmentDirective::create):
* Source/WebCore/page/FragmentDirective.idl: Added.

Canonical link: https://commits.webkit.org/279258@main
